### PR TITLE
image handler / update tests / add eslint file

### PR DESCRIPTION
**Pull request by RobertMrowiec**

### CODE REVIEW

File: carbid.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+require('./app').then(app => {
```
Comment (Rxbsxn) [inline carbid.js]: 🌮 👍

File: app.js [hunk 2]
```diff
@@ -17,8 +17,12 @@ app.use('/api/users', require('./routing/users/route'))
...
+app.use('/api/images', require('./routing/images/route'))
 app.use('/login', require('./routing/login/route'))
 
-return mongoose.connect('mongodb://localhost/carbid', { useNewUrlParser: true }).then(() => {
-    return app.listen(8007, () => console.log('Server is running on port: 8007'))
-})
\ No newline at end of file
+module.exports = () => {
+  return mongoose.connect(
+    process.env.MONGODB_URI || 'mongodb://localhost/carbid',
```
Comment (Rxbsxn) [inline app.js]: In your new approach, we can move setup of connection to the database to a separated file, what do you think?

File: routing/auctions/route.js [hunk 3]
```diff
@@ -1,10 +1,22 @@
 const express = require('express')
 const router = express.Router()
 const auctions = require('./details')
+const multer  = require('multer')
```
Comment (Rxbsxn) [inline routing/auctions/route.js]: 👍

File: routing/common.js [hunk 4]
```diff
@@ -1,7 +1,13 @@
 function defaultResponse(func) {
     return (req, res) => func(req, res)
         .then(result => res.status(200).json(result))
-        .catch(err => res.status(400).json(err))
+        .catch(err => {
+            if (err instanceof Error || err instanceof TypeError) {
+                console.log(err)
```
Comment (Rxbsxn) [inline routing/common.js]: 💇‍♂️

File: routing/images/route.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+const express = require('express')
+const router = express.Router()
+const images = require('./details')
+const multer  = require('multer')
+
+const storage = multer.diskStorage({
+    destination: function (req, file, cb) {
+      cb(null, 'uploads/')
+    },
```
Comment (Rxbsxn) [inline routing/images/route.js]: I see the same code above, am I right? I think we need to extract that functionality to separated file. What do you think?

File: routing/users/details.js [hunk 6]
```diff
@@ -15,12 +16,17 @@ exports.pagination = defaultResponse(req => {
...
-    
+
     return !result.length ? new User(req.body).save() : result
 })
 
 exports.update = defaultResponse(async req => {
+    req.body = req.body.data.attributes
+    const user = await User.findById(req.params.id)
     const result = await userValidate(req.body)
+    if (req.body.password && !bcrypt.compare(req.body.password, user.password)){
```
Comment (Rxbsxn) [inline routing/users/details.js]: Use destructuring for `req.body.password`

Comment (Rxbsxn) [reply routing/users/details.js]: `const { password } = req.body`

File: tests/auctions.test.js [hunk 7]
```diff
@@ -43,21 +44,21 @@ describe('GET auction by ID', () => {
 })
 
 describe('ADD auctions', () => {
-	test('add auctiion if body passes the validation', () => {
+	test('add auction if body passes the validation', () => {
+		
+		let form = new FormData()
```
Comment (Rxbsxn) [inline tests/auctions.test.js]: I'm a bit curious if cannot we just move the initialization of `FormData` instance to the top with these all properties and then modify them in specific test cases? The current approach is good 👍 , but I'm curious if we can be more DRY with mentioned approached

File: tests/token.js [hunk 8]
```diff
@@ -5,6 +5,6 @@ const dotenv = require('dotenv')
 dotenv.config()
 
 module.exports.generateToken = async () => {
-    const user = await new User({email: 'token@herecars.com', password: 'qwe123', name: 'token'}).save()
-    return jwt.sign({user: user._id}, process.env.privateKey)
+	const user = await new User({email: 'token@herecars.com', password: 'qwe123', name: 'token'}).save()
```
Comment (Rxbsxn) [inline tests/token.js]: :cotusie:

File: tests/users.test.js [hunk 9]
```diff
@@ -1,228 +1,193 @@
...
-        return fetch(`${url}/users/${userId}`, { headers: { 'Authorization': `Bearer ${token}`}})
-        .then(result => result.json())
-        .then(user => {
-            expect(200)
-            expect(user._id).toEqual(userId)
-        })
-    })
+	test('Get user by ID', () => {
+		request(app).get(`/api/users/${userId}`)
+			.set('Authorization', `Bearer ${token}`)
```
Comment (Rxbsxn) [inline tests/users.test.js]: ✌️

File: tests/auctions.test.js (hunk 7 above)
Comment (RobertMrowiec) [reply tests/auctions.test.js]: I tried to make only one assignment before I pushed these code, unlucky without success, but I will try ;)

File: tests/token.js (hunk 8 above)
Comment (RobertMrowiec) [reply tests/token.js]: INGLISZ PLIS  🏴󠁧󠁢󠁥󠁮󠁧󠁿

Comment (RobertMrowiec) [reply tests/token.js]: 🏴󠁧󠁢󠁥󠁮󠁧󠁿

Comment (Rxbsxn) [reply tests/token.js]: :gb:

File: tests/users.test.js (hunk 9 above)
Comment (RobertMrowiec) [reply tests/users.test.js]: everywhere is tab (2 spaces) :P

File: routing/images/route.js (hunk 5 above)
Comment (RobertMrowiec) [reply routing/images/route.js]: It will be changed soon cause right now it doesnt work as I expected

File: tests/users.test.js (hunk 9 above)
Comment (Rxbsxn) [reply tests/users.test.js]: I didn't complain about the indentations, I just like this test 🍺

File: routing/auctions/details.js [hunk 10]
```diff
@@ -23,14 +22,14 @@ exports.pagination = defaultResponse(async req => {
 })
 
 exports.add = defaultResponse(async req => {
-	const body = req.body.data.attributes
+	const body = { ...req.body.data.attributes, car: req.body.data.relationships.car.data.id }
```
Comment (Rxbsxn) [inline routing/auctions/details.js]: destruct

Comment (Rxbsxn) [reply routing/auctions/details.js]: { data } = req.body

File: routing/auctions/details.js [hunk 11]
```diff
@@ -22,7 +22,8 @@ exports.pagination = defaultResponse(async req => {
 })
 
 exports.add = defaultResponse(async req => {
-	const body = { ...req.body.data.attributes, car: req.body.data.relationships.car.data.id }
+	const { data } = req.body
+	const body = { ...data.attributes, car: data.relationships.car.data.id, user: data.relationships.car.data.id  }
```
Comment (Rxbsxn) [inline routing/auctions/details.js]: Are you sure about it? `user: data.relationship.car.data.id` + I really recommend destruct here on `relationship`

File: routing/auctions/details.js [hunk 12]
```diff
@@ -18,12 +18,16 @@ exports.pagination = defaultResponse(async req => {
...
-	return auctionSerialize(await Auction.find().sort('-endDate').skip(size * (number - 1)).limit(size), links, { 'total': totalPages })
+	return auctionSerialize(await Auction.find().sort('-endDate').populate('car').skip(size * (number - 1)).limit(size), links, { 'total': totalPages })
 })
 
+exports.byUser = defaultResponse(async req => auctionSerialize( await Auction.find({ user: req.params.userId}).exec()))
+
 exports.add = defaultResponse(async req => {
 	const { data } = req.body
-	const body = { ...data.attributes, car: data.relationships.car.data.id, user: data.relationships.car.data.id  }
+	const { relationships } = data
```
Comment (Rxbsxn) [inline routing/auctions/details.js]: Even we can do `const { relationships: { user, car } } = data`

File: routing/auctions/route.js [hunk 13]
```diff
@@ -15,6 +15,7 @@ const upload = multer({ storage: storage })
   
 router.get('/', auctions.pagination)
 	.get('/:id', auctions.getById)
+	.get('/byUser/:userId', auctions.byUser)
```
Comment (Rxbsxn) [inline routing/auctions/route.js]: `byUser`?

Comment (RobertMrowiec) [reply routing/auctions/route.js]: yest, /api/auctions/byUser/:userId . I think it makes logic here, but if not, just reply

File: routing/auctions/details.js (hunk 12 above)
Comment (RobertMrowiec) [reply routing/auctions/details.js]: Oh, didnt know that!

File: routing/auctions/route.js (hunk 13 above)
Comment (RobertMrowiec) [reply routing/auctions/route.js]: Ou, I see that in ember data I won't use rest like this, just with filter... It's going to be changed

File: routing/auctions/details.js [hunk 14]
```diff
@@ -7,9 +7,12 @@ exports.get = defaultResponse(() => Auction.find().populate('car').populate('off
 exports.getById = defaultResponse(async req => auctionSerialize( await Auction.findById(req.params.id).populate('car').populate('offer')))
 
 exports.pagination = defaultResponse(async req => {
+	const match = req.query.user ? { user: req.query.user } : {}
```
Comment (Rxbsxn) [inline routing/auctions/details.js]: please destruct user

File: routing/auctions/details.js [hunk 15]
```diff
@@ -18,16 +21,18 @@ exports.pagination = defaultResponse(async req => {
 		last: `${url(req)}/api/auctions?page[number]=${totalPages}&page[size]=${size}`
 	}
 
-	return auctionSerialize(await Auction.find().sort('-endDate').populate('car').skip(size * (number - 1)).limit(size), links, { 'total': totalPages })
+	if (req.query.user) for (x of Object.keys(links)) {
```
Comment (Rxbsxn) [inline routing/auctions/details.js]: Too much math? Rename x to something more meaningful

File: models/Auction.js [hunk 16]
```diff
@@ -4,7 +4,7 @@ const Schema = mongoose.Schema
 const AuctionSchema = new Schema({
 	car: { type: Schema.Types.ObjectId, ref: 'Car', required: true},
 	description: { type: String, required: true },
-	endDate: Date,
+	endDate: String,
```
Comment (Rxbsxn) [inline models/Auction.js]: Why `String` here?

File: models/Car.js [hunk 17]
```diff
@@ -2,7 +2,7 @@ const mongoose = require('mongoose')
 const Schema = mongoose.Schema
 
 const CarSchema = new Schema({
-	assembledAt: { type: Date, default: Date.now() },
+	assembledAt: String,
```
Comment (Rxbsxn) [inline models/Car.js]: same here

File: routing/auctions/details.js [hunk 18]
```diff
@@ -7,11 +7,12 @@ exports.get = defaultResponse(() => Auction.find().populate('car').populate('off
...
-	const match = req.query.user ? { user: req.query.user } : {}
-	const number = +req.query.page.number
-	const size = +req.query.page.size
+	const { userId, page } = req.query
+	const match = userId ? { user: userId } : {}
+	const number = +page.number
+	const size = +page.size
 
-	const totalPages = Math.ceil(await (req.query.user ? Auction.countDocuments(match) : Auction.countDocuments())/ size)
+	const totalPages = Math.ceil(await (userId ? Auction.countDocuments(match) : Auction.countDocuments()) / size)
```
Comment (Rxbsxn) [inline routing/auctions/details.js]: We can move this logic with await to a separated method for better readability

File: routing/auctions/details.js [hunk 19]
```diff
@@ -43,8 +41,23 @@ exports.add = defaultResponse(async req => {
...
-	const result = await auctionValidate(req.body)
-	return !result.length ? Auction.findByIdAndUpdate(req.params.id, req.body, {new: true}) : result
+	const body = setBody(req)
+
+	if (req.file) {
+		body.image = `http://localhost:8008/${req.file.path}`
+	}
+
+	const result = await auctionValidate(body)
+	return !result.length ? Auction.findByIdAndUpdate(req.params.id, body, {new: true}).then(data => auctionSerialize(data)) : result
```
Comment (Rxbsxn) [inline routing/auctions/details.js]: We can move the positive result of ternary to a separated method

File: routing/auctions/details.js [hunk 20]
```diff
@@ -43,8 +41,23 @@ exports.add = defaultResponse(async req => {
...
 })
 
-exports.delete = defaultResponse(req => Auction.findByIdAndDelete(req.params.id))
\ No newline at end of file
+exports.delete = defaultResponse(req => Auction.findByIdAndDelete(req.params.id))
+
+function setBody(req) {
```
Comment (Rxbsxn) [inline routing/auctions/details.js]: Maybe `_setBody`, because of this a private method somehow

File: routing/cars/details.js [hunk 21]
```diff
@@ -13,17 +13,23 @@ exports.pagination = defaultResponse(req => {
...
 	const result = await carValidate(req.body)
-	return !result.length ? Car.findByIdAndUpdate(req.params.id, req.body, {new: true}) : result
+	return !result.length ? Car.findByIdAndUpdate(req.params.id, req.body, {new: true}).then(data => carSerialize(data)) : result
 })
 
-exports.delete = defaultResponse(req => Car.findByIdAndDelete(req.params.id))
\ No newline at end of file
+exports.delete = defaultResponse(req => Car.findByIdAndDelete(req.params.id))
+
+function setBody(req) {
```
Comment (Rxbsxn) [inline routing/cars/details.js]: same here

File: models/Auction.js (hunk 16 above)
Comment (RobertMrowiec) [reply models/Auction.js]: due to format on front end

File: models/Car.js (hunk 17 above)
Comment (RobertMrowiec) [reply models/Car.js]: due to format on front end

**Comment by Rxbsxn** [comment]
@RobertMrowiec Shall we merge it or remove?

**Pull request closed without merging**